### ISSUE TITLE
Module parse failed

### ISSUE THREAD

**Issue by crassSandwich** [y-lohse/inkjs#427]
I get the following error when I try to use this package in a newly bootstrapped create-react-app application:

    ./node_modules/inkjs/dist/ink-es2015.js 491:14
    Module parse failed: Identifier 'e' has already been declared (491:14)
    File was processed with these loaders:
     * ./node_modules/babel-loader/lib/index.js
    You may need an additional loader to handle the result of these loaders.
    |           let t = _ref4[0];
    |           let e = _ref4[1];
    >           let e = f.fromSerializedKey(t);
    |           if (null === e.originName) return d("item.originName");
    |

Out of the versions I tested, it happens in 1.9.* and 1.10.* but not in 1.7 or 1.8.

Steps to reproduce:

1. make a project with create-react-app
2. cd into that project and install inkjs
3. modify App.js to import or require inkjs (don't have to use it anywhere, just have to try and load it)
4. `npm start`

I've only been playing around with create-react-app so I don't know if it's more general than that, but it does look like it might be an issue in any babel environment. Unfortunately I don't know enough about babel to set an example up.

**Comment by y-lohse** [comment]
Could you see if you can reproduce the problem on codesandbox or a similar service ? I just gave it a quick shot [here](https://codesandbox.io/s/dazzling-robinson-7qsvh) and it seems ok, but there could be a number of differences in setup.

**Comment by derekja** [comment]
I'm hitting this problem as well, I'm afraid. I cannot yet reproduce in the sandbox. I will keep trying on a repro. The commonalities with OP include using create-react-app as a starting point. More soon...

**Comment by derekja** [comment]
Slow progress on this one over the holidays. I still cannot repro in a sandbox. I strangely also cannot repro in a create-react-app that has been ejected. (The default create-react-app project has a lot of dependencies hidden from the user, and a one-way "eject" process makes all of those dependencies visible). Still looking at it some more, though!

**Comment by derekja** [comment]
Made a bit more headway on this, but not resolved yet. I see the issue reported by crassSandwich on version 1.9 (and above) but not version 1.8. When I take just the ink-es2015.js file in to the version of inkjs in node_modules I can get it to break (or not break, depending on the version I move in there.)

When I look at the changes between 1.8 and 1.9 there are a number, but one change in particular stands out. When I change the rollup config in 1.9 to omit the terser plugin everything works fine. Correspondingly, when I insert the terser plugin into 1.8 I get exactly the error reported at the top of this thread.

I'm learning a lot about webpack and rollup through this investigation, so I'll keep digging. I'm still not sure why I see this issue only with create-react-app.

**Comment by derekja** [comment]
OK, I'm going to have to stop on this. Vacation is over and I have a workaround for my needs (which unfortunately means building a modified version of inkjs without the terser plugin) but still can't quite figure out why it only happens with create-react-app projects. Options looked at and excluded:

-- since webpack runs terser for it's minification, maybe running terser on a file that has already been through terser causes problems. Nope, webpack terser is only run on production builds and the problem also shows up in dev builds.
-- maybe it actually has more to do with the sourcemaps that terser automatically produces than it does the minified code. Nope - changing the default output of sourcemaps doesn't help. Can't find a set of terser options that impact the behaviour at all actually.
-- I hand-built a babel, webpack, react sample and it did not have the issue. I started moving in code from the ejected create-react-app environment, but couldn't find a minimal combination that showed where the problem was coming from. Stopped because having to eject all create-react-app projects to change their config isn't a viable solution anyway.

Please let me know if anyone else picks this up or has anything I can shed light on.

Thanks!

**Comment by y-lohse** [comment]
Thanks for the investigation and the write-up!

**Comment by carlwells** [comment]
Do you actually need to `require inkjs from 'inkjs';`? This causes the same error for me.
However…
Docs state: Require the module: `var Story = require('inkjs').Story;`
As does @y-lohse above sandbox example using `import { Story } from "inkjs";`

Importing the Story module doesn't throw an error. I haven't yet integrated any further though so I can't comment on any other issue.

**Comment by derekja** [comment]
Thanks for the thought! I got excited for a moment thinking I had just missed something silly. Unfortunately, I can confirm that (at least on my system) starting a new create-react-app project and adding the latest inkjs from npm and in the generated App.js adding the line "import { Story } from 'inkjs';" results in the error at the top of this thread.

**Comment by kris10cabrera** [comment]
I'm also hitting this problem.

**Comment by kris10cabrera** [comment]
@derekja would you mind sharing your modified version of inkjs without the terser plugin?

**Comment by derekja** [comment]
Oh sure, in rollup.config.js just comment out line 52.

I actually decided that I wanted to know more about how create-react-app was working, though, so I ejected and customized my webpack and start scripts to not use the ones provided by create-react-app. So I guess in my case my workaround finally ended up being to not use create-react-app. I never did figure out exactly where the conflict with the terser plugin was arising, though.

**Comment by kris10cabrera** [comment]
Thank you! @derekja

**Comment by carlwells** [comment]
I'm hitting this problem, I thought it was working @derekja but I think _I_ was doing something silly.

I think I have found where the issue occurs in the ink build.

The terser plugin keeps the unused variable `value` in the for loop in inkjs-master/src/InkList.ts. Ln224 Col 24
```
for (let [key, value] of this) {
  let item = InkListItem.fromSerializedKey(key);
  if (item.originName === null) return throwNullException('item.originName');
  this._originNames.push(item.originName);
}
```
This gets output as in inkjs-es2015.js as: _(formatted for clarity)_
```
for(let[t,e]of this){
  let e=f.fromSerializedKey(t);
  if(null===e.originName) return d("item.originName");
  this._originNames.push(e.originName)
}
```
so webpack complains that `Identifier 'e' has already been declared` because`let e=f.fromSerializedKey(t);` is trying to overwrite the `e` value from the loop.

Since the `e` value in the loop is unused I've found that I can safely remove it from InkList.ts and rebuild. This prevents the issue.

I've tries various terser config options but I'm unable to get it to remove the unused variable `e` from the for loop.

@y-lohse I can make a PR if you want and you can take a look?

**Comment by y-lohse** [comment]
@carlwells that looks like a terser bug..? But anyway, nice find and yeah I'd be happy to accept a PR with this fix :)

**Comment by lina94** [comment]
I'm just starting with an Angular app and I'm getting a similar error on build, coming from babel crawl and traverse functions: `Duplicate declaration "e"`

I could replicate this when pasting the ink-es2015.js in the [Babel Test Site](https://babeljs.io/repl) with es2015. 

Do you think it's the same issue? Are there plans for fixing this or could you suggest something I can try as a workaround? Thanks! :)

**Comment by y-lohse** [comment]
It's likely the same issue, yes.

I have no plans for a fix at the moment, though I'm happy to test and look at pull requests. Maybe the bug in terser has been resolved since then, so one thing that you can try is to build inkjs from master and see if the bug is still there. If not let me know and I can release a new version.
Otherwise, you could try to use the [non-es2015 version](https://unpkg.com/browse/inkjs@1.10.3/dist/ink.js) which might be fine?

**Comment by lina94** [comment]
Thanks @y-lohse ! The following workaround worked for me: 
I removed the es2015 version from the node module and changed the package.json of the module so that "main": "dist/ink.js" (instead of the es2015 version)

**Comment by y-lohse** [comment]
Cool! Note that you can probably do something like `import Story from 'inkjs/dist/ink.js'` without changing the dependency itself, though it depends how your bundler works. But glad you found a workaround :)

**Comment by y-lohse** [comment]
Continuing #466: thanks @furkle for the extra details, especially the fact that the error doesn't happen in dev mode. This is why it worked on my codesandbox example, but if I try to deploy that example the build does fail.

I'm taking a look at the solutions suggested by @carlwells .

**Comment by y-lohse** [comment]
I just released version 1.10.4, it solves the duplicate `e` declaration mentioned by @carlwells but it looks like there might be others in there. Wouldn't mind a confirmation that the issue is still present though.

**Comment by y-lohse** [comment]
I'm fairly confident that version 1.11.0 doens't have this problem anymore, so I'll close this issue. Feel free to reopen it if you run into the same problem!
